### PR TITLE
Add an admin request to return the persistent index values for a blob id

### PR DESCRIPTION
**Pull request by justinlin-linkedin**
The issue this PR is trying  to solve, is “How do we know if a given blob exist in a given server?”

Before this PR, we have to copy all the index segment files to local machine and then run DumpIndexTool cli to print out all the blobs from the index segment files that we copied. This is of course too much work. 

With this PR, we can just send a request to this host and get the result right away, assuming this host is up and running.

### CODE REVIEW

**Review by snalli — COMMENTED** [review]
Some context as to why you are making all these changes would be relevant. Is this for a GCN ? If so, just paste the GCN number in the commit msg.

File: ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java [hunk 1]
```diff
@@ -493,10 +505,57 @@ private AdminResponse handleHealthCheckRequest(DataInputStream requestStream, Ad
...
+    try {
+      PartitionId partitionId = blobId.getPartition();
+      Store store = storeManager.getStore(partitionId);
+      if (store == null) {
+        logger.error("BlobId {}'s partition[{}] doesn't exist in this host", blobId, partitionId.getId());
+        return new AdminResponseWithContent(correlationId, clientId, ServerErrorCode.Partition_Not_Here, null);
+      }
+
+      List<MessageInfo> messages = store.findAllMessageInfoForKey(blobId);
+      Collections.sort(messages, (m1, m2) -> (int) (m1.getOperationTimeMs() - m2.getOperationTimeMs()));
```
Comment (snalli) [inline ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java:546]: Why is this sort needed ?

File: ambry-protocol/src/main/java/com/github/ambry/protocol/BlobIndexAdminRequest.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  /**
+   * Constructor for {@link BlobIndexAdminRequest}.
+   * @param blobId The blob id in string format.
+   * @param adminRequest the {@link AdminRequest} that contains common admin request related information.
+   */
+  public BlobIndexAdminRequest(String blobId, AdminRequest adminRequest) {
+    super(AdminRequestOrResponseType.BlobIndex, adminRequest.getPartitionId(), adminRequest.getCorrelationId(),
+        adminRequest.getClientId());
+    this.blobId = blobId;
+    this.sizeInBytes = super.sizeInBytes() + Short.BYTES + BLOB_ID_LENGTH + blobId.length();
```
Comment (snalli) [inline ambry-protocol/src/main/java/com/github/ambry/protocol/BlobIndexAdminRequest.java:56]: plus the Long.BYTES for sizeInBytes ?

File: ambry-protocol/src/main/java/com/github/ambry/protocol/BlobIndexAdminRequest.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  @Override
+  public long sizeInBytes() {
+    return sizeInBytes;
+  }
+
+  @Override
+  public String toString() {
+    return "BlobIndexAdminRequest[ClientId=" + clientId + ", CorrelationId=" + correlationId + ", BlobId=" + blobId
```
Comment (snalli) [inline ambry-protocol/src/main/java/com/github/ambry/protocol/BlobIndexAdminRequest.java:73]: String result;
result += "BlobIndexAdminRequest[ClientId=" + clientId;
result += ", CorrelationId=" + correlationId;
...

Just more readable

File: ambry-protocol/src/test/java/com/github/ambry/protocol/RequestResponseTest.java [hunk 4]
```diff
@@ -949,13 +947,16 @@ public void replicationControlAdminRequestTest() throws IOException {
...
+   * @throws IOException
+   */
   @Test
   public void healthCheckAdminRequestTest() throws IOException {
     MockClusterMap clusterMap = new MockClusterMap();
     int correlationId = 1;
     String clientId = "ambry-healthchecker";
-    AdminRequest adminRequest =
-        new AdminRequest(AdminRequestOrResponseType.RequestControl, null, correlationId, clientId);
+    AdminRequest adminRequest = new AdminRequest(AdminRequestOrResponseType.HealthCheck, null, correlationId, clientId);
```
Comment (snalli) [inline ambry-protocol/src/test/java/com/github/ambry/protocol/RequestResponseTest.java:959]: is this part of this PR ? if not, can u make a separate PR ?

File: ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java [hunk 5]
```diff
@@ -493,10 +509,59 @@ private AdminResponse handleHealthCheckRequest(DataInputStream requestStream, Ad
...
+    } catch (IOException e) {
+      logger.error("Failed to convert {} to a BlobId", blobIndexAdminRequest.getBlobId(), e);
+      return new AdminResponseWithContent(correlationId, clientId, ServerErrorCode.Bad_Request, null);
+    }
+
+    try {
+      PartitionId partitionId = blobId.getPartition();
+      Store store = storeManager.getStore(partitionId);
+      if (store == null) {
+        logger.error("BlobId {}'s partition[{}] doesn't exist in this host", blobId, partitionId.getId());
```
Comment (snalli) [inline ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java:546]: BlobId {}'s -> BlobId[{}]

File: ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java [hunk 6]
```diff
@@ -493,10 +509,59 @@ private AdminResponse handleHealthCheckRequest(DataInputStream requestStream, Ad
...
+      // Serialize with JSON
+      String json = objectMapper.writeValueAsString(messages);
+      return new AdminResponseWithContent(correlationId, clientId, ServerErrorCode.No_Error, json.getBytes());
+    } catch (StoreException e) {
+      logger.error("Failed to find all message infos for given key {}", blobId, e);
+      errorMessage = e.getMessage();
+    } catch (IOException e) {
+      logger.error("Failed to serialize to json string", e);
+      errorMessage = e.getMessage();
+    }
```
Comment (snalli) [inline ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java:561]: Can there be an uncaught exception here ?

File: ambry-store/src/main/java/com/github/ambry/store/BlobStore.java [hunk 7]
```diff
@@ -957,6 +958,30 @@ public MessageInfo findKey(StoreKey key) throws StoreException {
...
+  public List<MessageInfo> findAllMessageInfoForKey(StoreKey key) throws StoreException {
+    checkStarted();
+    final Timer.Context context = metrics.findKeyResponse.time();
+    try {
+      List<IndexValue> values = index.findAllIndexValuesForKey(key, null);
+      if (values == null) {
+        return Collections.emptyList();
+      }
+      return values.stream()
+          .map(value -> new MessageInfo(key, value.getSize(), value.isDelete(), value.isTtlUpdate(), value.isUndelete(),
```
Comment (snalli) [inline ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:971]: how about create a new ctor like MessageInfo(key, value) ?

**Comment by snalli** [comment]
Can u break this PR down ? Its too big

**Review by snalli — CHANGES_REQUESTED** [review]
requesting changes

**Review by JingQianCloud — COMMENTED** [review]
Nice tool which I'd like to have.
It's be helpful for me to test On-demand replication as well.

Just a feature requirement,
current the implementation is based on PersistentIndex.findAllIndexValuesForKey which returns List of IndexValue.
I wish I can get not only the IndexValues but also the Index Segment/Log Segment file names the IndexValues reside.
Wish we can have it in the next revision.

File: ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java [hunk 8]
```diff
@@ -493,10 +509,60 @@ private AdminResponse handleHealthCheckRequest(DataInputStream requestStream, Ad
...
+      logger.error("Failed to deserialize BlobIndexAdminRequest", e);
+      return new AdminResponseWithContent(correlationId, clientId, ServerErrorCode.Bad_Request, null);
+    }
+
+    BlobId blobId = null;
+    try {
+      String blobIdStr = blobIndexAdminRequest.getBlobId();
+      blobId = new BlobId(blobIdStr, clusterMap);
+    } catch (Exception e) {
+      logger.error("Failed to convert {} to a BlobId", blobIndexAdminRequest.getBlobId(), e);
```
Comment (tbingol) [inline ambry-server/src/main/java/com/github/ambry/server/AmbryServerRequests.java:538]: nit: s/blobIndexAdminRequest.getBlobId/blobIdStr

**Review by tbingol — APPROVED** [review]

**Review by ankagrawal — COMMENTED** [review]
It looks like some of the changes of this PR are present in other PR. Are we planning to divide this PR into smaller PRs?

**Comment by justinlin-linkedin** [comment]
> It looks like some of the changes of this PR are present in other PR. Are we planning to divide this PR into smaller PRs?

that's right, I will close this one.

**Pull request closed without merging**